### PR TITLE
Dash svm

### PR DESCRIPTION
**Pull request by YunkeXiao**
Issue for app: #39 

# App pull request
- [ ] This is a new app
- [x] I am improving an existing app (redesigns/code "makeovers")

## About

- Playground deployment URL (new version): https://dash-playground.plotly.host/dash-svm/
- Current gallery app URL: https://dash-gallery.plotly.host/dash-svm/

## Workflow

- [x] I have created a branch in the appropriate monorepo, and the
  elements necessary for successful deployment are in place.
- [x] If the app is a redesigned and/or restyled version of an
  existing gallery app, I've summarized the changes requested in the
  appropriate Streambed issue and confirm that they have been applied.
- [x] If the app is on the Dash Gallery portal, I have added a link to
  the GitHub repository for the source code in the portal description.
- [x] If the app is a reimplementation of a Python gallery app for the
  DashR gallery, the app in this PR mimics, as closely as possible,
  the style and functionality of the existing app.

## The pre-review review

I have addressed all of the following questions:

- [x] Does everything in my code serve some purpose? (I have removed
  any dead and/or irrelevant code.)
- [x] Does everything in my code have a clear purpose? (My code is
  readable and, where it isn't, it has been commented appropriately.)]
- [x] Am I reinventing the wheel? (I have used appropriate packages to
  lessen the volume of code that needs to be maintained.)

### CODE REVIEW

File: apps/dash-svm/CONTRIBUTING.md [hunk 1]
```diff
@@ -0,0 +1,92 @@
+# Contributing
```
Comment (shammamah) [inline apps/dash-svm/CONTRIBUTING.md]: None of the other apps have this, let's get rid of it.

File: apps/dash-svm/LICENSE.md [hunk 2]
```diff
@@ -0,0 +1,21 @@
+The MIT License (MIT)
```
Comment (shammamah) [inline apps/dash-svm/LICENSE.md]: You can also get rid of this file.

File: apps/dash-svm/README.md [hunk 3]
```diff
@@ -0,0 +1,82 @@
+# Support Vector Machine (SVM) Explorer [![GitHub license](https://img.shields.io/github/license/plotly/dash-svm.svg)](https://github.com/plotly/dash-svm/blob/master/LICENSE.md) [![Mentioned in Awesome Machine Learning](https://awesome.re/mentioned-badge.svg)](https://github.com/josephmisiti/awesome-machine-learning)
```
Comment (shammamah) [inline apps/dash-svm/README.md]:
```suggestion
# Support Vector Machine (SVM) Explorer [![Mentioned in Awesome Machine Learning](https://awesome.re/mentioned-badge.svg)](https://github.com/josephmisiti/awesome-machine-learning)
```

File: apps/dash-svm/requirements.txt [hunk 4]
```diff
@@ -0,0 +1,15 @@
+# Core
+gunicorn>=19.8.1
+plotly==2.7.0
+dash>=0.40.0
+dash-renderer>=0.21.0
```
Comment (shammamah) [inline apps/dash-svm/requirements.txt]: You don't need renderer, html, or core any more (they're auto-installed with Dash).

File: apps/dash-svm/requirements.txt [hunk 5]
```diff
@@ -0,0 +1,15 @@
+# Core
+gunicorn>=19.8.1
+plotly==2.7.0
+dash>=0.40.0
```
Comment (shammamah) [inline apps/dash-svm/requirements.txt]: Please upgrade to latest version of Dash!

File: apps/dash-svm/README.md (hunk 3 above)
Comment (shammamah) [reply apps/dash-svm/README.md]: @YunkeXiao You didn't actually make this change, please commit the suggestion above.

**Comment by shammamah** [comment]
@YunkeXiao All the CSS appears to be gone on the playground deployment -- can you please make sure that your assets path is correct?

File: apps/dash-svm/assets/custom-styles.css [hunk 6]
```diff
@@ -0,0 +1,266 @@
...
+.rc-slider-track {
+    background-color: #13c6e9 !important;
+}
+
+.rc-slider-handle {
+    border: solid 2px #13c6e9 !important
+}
+
+input[type='radio']:checked {
+    background-color: #13c6e9 !important;
```
Comment (Mskycoder) [inline apps/dash-svm/assets/custom-styles.css]: Are these `!important` flags absolutely necessary? It's a nuclear last resort solution, so we should use built-in precedence rules whenever possible.

File: apps/dash-svm/app.py [hunk 7]
```diff
@@ -0,0 +1,443 @@
...
+
+    roc_figure = figs.serve_roc_curve(model=clf, X_test=X_test, y_test=y_test)
+
+    confusion_figure = figs.serve_pie_confusion_matrix(
+        model=clf, X_test=X_test, y_test=y_test, Z=Z, threshold=threshold
+    )
+
+    print(f"Total Time Taken: {time.time() - t_start:.3f} sec")
+
+    return [
```
Comment (Mskycoder) [inline apps/dash-svm/app.py]: How significant is the performance impact of returning large layout elements? Would using an intermediate value to store the trained SVMs locally reduce the load on the network and reduce latency?

File: apps/dash-svm/assets/custom-styles.css [hunk 8]
```diff
@@ -0,0 +1,266 @@
...
+}
+
+/* Left column
+–––––––––––––––––––––––––––––––––––––––––––––––––– */
+#slider-svm-parameter-C-coef {
+    padding: 5px 10px 25px;
+}
+
+#slider-svm-parameter-gamma-coef {
+    padding: 5px 10px 25px
```
Comment (Mskycoder) [inline apps/dash-svm/assets/custom-styles.css]: Adding a semi-colon at the end of each rule is better practice since it's improves maintainability of the code even if there are no significant performance impact.

File: apps/dash-svm/app.py (hunk 7 above)
Comment (YunkeXiao) [reply apps/dash-svm/app.py]: That was actually code that was used the original author of the application, so I think it's safe to assume that was already taken into account and that the current state of the application is somewhat optimal.

File: apps/dash-svm/assets/custom-styles.css (hunk 6 above)
Comment (YunkeXiao) [reply apps/dash-svm/assets/custom-styles.css]: Some of these are necessary if we're to avoid inline styling within components.

File: apps/dash-stitching/README.md [hunk 9]
```diff
@@ -0,0 +1,47 @@
+# Dash Image Stitching and Registration
```
Comment (shammamah) [inline apps/dash-stitching/README.md]: Why are these files in here? This isn't the stitching app.

File: apps/dash-svm/app.py [hunk 10]
```diff
@@ -0,0 +1,441 @@
...
+
+        rng = np.random.RandomState(2)
+        X += noise * rng.uniform(size=X.shape)
+        linearly_separable = (X, y)
+
+        return linearly_separable
+
+    else:
+        raise ValueError(
+            "Data type incorrectly specified. Please choose an existing " "dataset."
```
Comment (shammamah) [inline apps/dash-svm/app.py]: Could you get rid of these extra quotation marks?

File: apps/dash-svm/app.py (hunk 7 above)
Comment (shammamah) [reply apps/dash-svm/app.py]: It's possible that there could be some performance improvements with what @Mskycoder suggested! Let's leave that particular refactoring out of this PR and maybe focus on improving the performance in another one.

File: apps/dash-svm/requirements.txt [hunk 11]
```diff
@@ -0,0 +1,10 @@
+# Core
+gunicorn>=19.8.1
+dash>=1.0.0
```
Comment (shammamah) [inline apps/dash-svm/requirements.txt]: Yay 🎉

**Comment by shammamah** [comment]
The logo on the app looks way too big - please make it a bit smaller! Also if possible, wrap the graphs in a `dcc.Loading` component - sometimes it's hard to tell if the app is broken or if it's just "thinking".

File: apps/dash-svm/app.py [hunk 12]
```diff
@@ -425,13 +425,27 @@ def update_svm_graph(
...
         ),
         html.Div(
             id="graphs-container",
             children=[
-                dcc.Graph(id="graph-line-roc-curve", figure=roc_figure),
-                dcc.Graph(id="graph-pie-confusion-matrix", figure=confusion_figure),
+                dcc.Loading(
+                    className="graph-wrapper",
+                    children=dcc.Graph(id="graph-line-roc-curve", figure=roc_figure),
+                    style={"display": "none"},
```
Comment (shammamah) [inline apps/dash-svm/app.py]: Why is this here? I think it will just hide the loading indicator, which is not what we want (we want for the app to show when it's loading something).

**Comment by shammamah** [comment]
@YunkeXiao I'm getting an application error when I try to visit the playground app. Please fix it, then I can review :)

File: apps/dash-svm/assets/custom-styles.css [hunk 13]
```diff
@@ -1,5 +1,7 @@
 @import url('https://fonts.googleapis.com/css?family=Playfair+Display');
 
+/* Body
```
Comment (shammamah) [inline apps/dash-svm/assets/custom-styles.css]: You can get rid of this comment.

File: apps/dash-svm/assets/custom-styles.css [hunk 14]
```diff
@@ -0,0 +1,450 @@
...
+
+    #graphs-container .graph-wrapper:nth-of-type(2) {
+        height: 40rem;
+        padding-left: 3rem;
+    }
+}
+
+/* Remove Undo
+–––––––––––––––––––––––––––––––––––––––––––––––––– */
+._dash-undo-redo {
```
Comment (celinehuang) [inline apps/dash-svm/assets/custom-styles.css]: remove this since the new version of Dash does this for us!
